zerver/tests/test_auth_backends.py

```diff
@@ -1471,6 +1476,305 @@ def test_social_auth_invited_as_admin_but_expired(self) -> None:
+                                                       name='King Hamlet')
+        action_key = create_email_change_confirmation_key(user_profile, new_email)
+
+        do_set_realm_property(user_profile.realm, 'email_changes_disabled', True)
+        self.login_user(user_profile)
+        result = self.social_auth_test(account_data_dict, subdomain='zulip',
+                                       action_key=action_key)
+
+        self.assertEqual(result.status_code, 302)
+        data = load_subdomain_token(result)
+        self.assertEqual(data['email'], self.example_email("hamlet"))
+        self.assertEqual(data['full_name'], self.name)
+        self.assertEqual(data['subdomain'], 'zulip')
+        self.assertEqual(data['action_key'], action_key)
+
+        result = self.client_get(result.url)
+        self.assertEqual(result.status_code, 400)
+        self.assert_in_response("Email address changes are disabled in this organization.",
+                                result)
+
+    def test_social_auth_email_change_delivery_email_with_admins_visibility_logged_in(self) -> None:
+        user_profile = self.example_user('hamlet')
+        new_email = 'hamlet-new@zulip.com'
+        old_email = user_profile.delivery_email
+        account_data_dict = self.get_account_data_dict(email=user_profile.delivery_email,
+                                                       name='King Hamlet')
+        action_key = create_email_change_confirmation_key(user_profile, new_email,
+                                                          Realm.EMAIL_ADDRESS_VISIBILITY_ADMINS)
+
+        self.login_user(user_profile)
+        result = self.social_auth_test(account_data_dict, subdomain='zulip',
+                                       action_key=action_key)
+
+        self.assertEqual(result.status_code, 302)
+        data = load_subdomain_token(result)
+        self.assertEqual(data['email'], self.example_email("hamlet"))
+        self.assertEqual(data['full_name'], self.name)
+        self.assertEqual(data['subdomain'], 'zulip')
+        self.assertEqual(data['action_key'], action_key)
+
+        result = self.client_get(result.url)
+        self.assert_in_success_response(["This confirms that the email address for your Zulip"],
+                                        result)
+        user_profile.refresh_from_db()
+        user_profile = get_user_by_delivery_email(new_email, user_profile.realm)
+        self.assertTrue(bool(user_profile))
+        user_profile = get_user_profile_by_id(user_profile.id)
+        self.assertEqual(user_profile.email, f"user{user_profile.id}@zulip.testserver")
+        with self.assertRaises(UserProfile.DoesNotExist):
+            get_user_by_delivery_email(old_email, user_profile.realm)
+
+        email_change_object = get_object_from_key(action_key, Confirmation.EMAIL_CHANGE)
+        email_change_object.refresh_from_db()
+        self.assertEqual(email_change_object.status, 1)
```

Comment:
This ratjer repetetive logic across the tests should probably be in some kind of `verify_user_email_changed` helper. I see that this is how the old email change tests were already done, but I don't understand why this validation is done in such a roundabout way - wouldn't it be sufficient to check `user_profile.delivery_email` is the correct new value, and that `status` of the email change object is changed and that's all? :thinking:

Reply:
Hmm, I guess checking if there's no account exists with the old email is worthy. The assertion to the non-delivery email is only done in `test_social_auth_email_change_delivery_email_with_admins_visibility` which I think is the whole point of the test. 

`verify_user_email_changed` sounds like a good idea. But I'm confused where to put that function in as it needs to be in a test subclassing `TestCase` for us to be able to use `self.assertEqual` and similar functions and we have 3 places(3 different test classes) we want to use it. I don't think it's a good idea to put it in `ZulipTestCase`. 

What do you think would be a good solution for this so that we can avoid duplication?

Reply:
Hmm, why would it be bad to put it in zuliptestcase?

Reply:
We would use `verify_user_email_changed` for just a very few tests cases and `ZulipTestCase` is being subclassed for every test class in our test code and most of them do not need it. It didn't feel like a good idea to me to put a function that has no use in most of the subclasses it stays hidden.
- - -
Now I re-wrote all these tests into mixin class and only the part that differs in each flow is extracted into a different function(`do_email_change`) which needs to be defined in the classes implementing this mixin. This reduced the amount of test code to one-third of what we had earlier.